### ISSUE TITLE
Add additional regression (or classification) head to the model

### ISSUE THREAD

**Issue by longxvu** [ultralytics/yolov5#2138]
## ❔Question

How to add additional regression (or classification) head to the model? Should I make change to the yolov5.yaml or should I make change to the Detect module? Is it as simple as changing 
```
self.no = nc + 5 
#to
self.no = nc + 7
```
https://github.com/ultralytics/yolov5/blob/73a066993051339f6adfe5095a7852a2b9184c16/models/yolo.py#L28-L32

How do I go about formatting the label files to reflex this change?
Thank you for your help

**Comment by glenn-jocher** [comment]
@longxvu `self.no` (number of outputs) is computed automatically, you should not change it. You can see examples of YOLOv5 with different numbers of detection heads:

- yolov5l.yaml has P3-P5 detection heads
- yolov5l6.yaml has P3-P6 detection heads
- yolov5-p2.yaml has P2-P5 detection heads

You can find all of these files in models and models/hub, and you can use them as guidance for formatting your own custom changes, i.e.:
https://github.com/ultralytics/yolov5/blob/master/models/hub/yolov5l6.yaml

**Comment by longxvu** [comment]
I think you misunderstood my question. If I understand correctly about P detection heads, they represent the stride apply to convolutional network. P2-P6 detection heads helps with detecting objects of different sizes (smaller object -> extra large object).

I will clarify my question further: For normal YOLO output, the output size for each detection is (5 + 80) (5: (x, y, w, h, conf) and 80 is the number of classes). However, for each object, I want to predict one more variable b, so the output will be (x, y, w, h, b, conf) + 80 classes. How should I go about adding 1 more variable to predict to the existing model?
Thank you.

**Comment by glenn-jocher** [comment]
@longxvu oh yes, you can add additional output neurons to train on additional label dimensions (like rotation for example). You should be able to update `shape.no` with this info as in your original post, and the Detect() forward function to handle the additional values properly.

If the additional dimensions are trainable and have loss, you'll also want to modify the 1) labels, 2) dataloader, and 3) loss function to be able to train on them.

**Comment by segalinc** [comment]
@longxvu how are you preparing your data in this case? I am facing the same problem

**Comment by longxvu** [comment]
> @longxvu how are you preparing your data in this case? I am facing the same problem

To prepare data what I did is to follow this link first: https://docs.ultralytics.com/yolov5/tutorials/train_custom_data
Then add any additional data columns that your problem need (cls x y w h feature1 feature2) instead of (cls x y w h).
Then change how the data is loaded as label here: https://github.com/ultralytics/yolov5/blob/3f74cd9ed1a17de94ed4a19dcefbaa3b27d16a95/utils/datasets.py#L455 (check the assert statement). Modify the function as appropriate to fit your need

**Comment by thohemp** [comment]
@longxvu Could you provide more information about your solution of implementing a regression variable?

**Comment by jobeid1** [comment]
@longxvu I am attempting to add an additional regression neuron to the output. I believe my labeling/dataloading is fine as you described above. Any advice or example on how you changed the Detect() function and the loss function? My loss function will be in the form: abs(actual-predicted)/actual. I'm very new to this and Python so any help would be useful!

**Comment by Light--** [comment]
same problem here, only add one neuron to self.no, but the trained model performs badly....

**Comment by Hasan-UTK18** [comment]
@longxvu were you succeeded to add the regression output layer? I am in the same problem. Any help would be appreciated.

**Comment by assafkeller** [comment]
@longxvu did you do it? 
can you share your code, please?

**Comment by assafkeller** [comment]
@glenn-jocher 
is the anchor needs to be changed? 
where else in the code is needed to be changed?
